docs/pbc/all/order-management-system/202311.0/base-shop/datapayload-conversion/state-machine/common-pitfalls-in-oms-design.md

```diff
@@ -154,6 +168,83 @@ The unused state could have a missing transition.
+
+**Solution:** In the described scenario, you probably don’t need the final state closed at all. The best approach would be to replace timeout with a manual trigger if an automatic transition is needed. Try to avoid timeouts with a duration of more than 7 days.
+
+## Long chain onEnter
+
+**Example:** When recalculation is started by oms:check-condition it will trigger a chain of onEnter events with more than 8 transitions in it.
+![long-chain-on-enter](https://github.com/xartsiomx/spryker-docs/assets/110463030/2cafb0fe-1388-434e-a783-7838535a69e7)
+
+**The issue:** Long chains onEnter events can be “fragile”. It will increase the time of execution, memory consumption as well as the risk of having an error in the middle of the process and ending up with a stuck order item.
+
+**Solution:** Remove unnecessary states/onEnter transition in your process. If you have an event with onEnter that doesn’t have any command/condition - probably it’s redundant. Split long chains into several smaller ones. Especially for chains that are starting from check* (condition, timeout) commands.
+
+## Slow order creation
+
+**The issue:** Order starting processing directly after placement. The checkout endpoint contains the execution of all onEnter transitions in OMS.
+![slow-order-creation](https://github.com/xartsiomx/spryker-docs/assets/110463030/02892077-3d8d-432e-a6a4-281dcdb9824d)
+
+**Solution:** Checkout endpoint logic should finish after the order was created with all items in starting states(new for example). The transition from the new state shouldn’t have an event (Which means it will be processed by oms:check-condition command)
+
+## Stuck on-enter
+
+**The issue:** In most cases, if you have an order item stuck during the onEnter transition it means you have an unexpected error during execution. I highly recommend checking those issues individually and properly fixing them.
+
+**Solution:** (if you have many of them due to infra, timeouts etc. and you need to fix them ASAP). Create a console command that will trigger onEnter events. I recommend creating a list of OMS states from which you want to check and move order items. Also, set limits for orders (not items!) and time windows. For example, check stuck order items only if they have the last update of the state after 2 hours.
+
+## Saving states (per item transaction)
+
+**Example:** The system has a callback that should move order items from picking started to ready for recalculation. After that, check-condition will move the order further to the recalculation step.
+![saving-states-1](https://github.com/xartsiomx/spryker-docs/assets/110463030/dd060438-7cd2-4453-9cf4-e08cda243003)
```

Comment:
Any chance for a screenshot with a straight arrow and some less aggressive color? :D